h5io/tests/test_io.py

```diff
@@ -499,3 +500,105 @@ def test_state_python_version_error(tmp_path):
         overwrite="update",
         use_state=True,
     )
+
+
+class Singleton(ABCMeta):
+    """
+    Copied from pyiron_base.
+
+    https://github.com/pyiron/pyiron_base/blob/33910343e5e6d4c8bbb5f2522ad6714ec5184ff5/pyiron_base/interfaces/singleton.py#L23
+    """
+
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        """If an instance already exists, return that one instead."""
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class RawSingleton(metaclass=Singleton):
+    """A toy singleton that defines `__reduce__` to de)serialize an associated var."""
+
+    def __reduce__(self):
+        """Return a string associated with the local instance of this class."""
+        return "raw"  # The associated global variable
+
+
+raw = RawSingleton()
+
+
+class UpdatingSingleton(metaclass=Singleton):
+    """A toy singleton that preserves data added between save and load."""
+
+    def __reduce__(self):
+        """Return a string associated with the local instance of this class."""
+        return "updating"
+
+    def __setstate__(self, state):
+        """Don't just set it, but update it in case more was added."""
+        self.__dict__.update(**state)
+
+
+updating = UpdatingSingleton()
+
+
+@pytest.mark.skipif(sys.version_info < (3, 11), reason="requires python3.11 or higher")
+def test_state_with_singleton(tmp_path):
+    """When __reduce__ returns a string, load the identical object."""
+    test_file = tmp_path / "test.hdf5"
+
+    raw.foo = "foo"
+
+    write_hdf5(
+        fname=test_file,
+        data=raw,
+        overwrite=True,
+        use_json=False,
+        use_state=True,
+    )
+
+    raw.bar = "bar"
+    raw_reloaded = read_hdf5(fname=test_file)
+
+    assert raw_reloaded is raw
+    assert raw_reloaded.foo == "foo"
+
+    # Reloaded uses __setstate__, which overwrites the __dict__ and we expect to lose
+    # state data added to the singleton between saving and loading
+    pytest.raises(
+        AttributeError,
+        getattr,
+        raw,
+        "bar",
+    )
+    pytest.raises(
+        AttributeError,
+        getattr,
+        raw,
+        "bar",
+    )
+
+    updating.foo = "foo"
+
+    write_hdf5(
+        fname=test_file,
+        data=updating,
+        overwrite=True,
+        use_json=False,
+        use_state=True,
+    )
+
+    updating.bar = "bar"
+    updating.foo = "not foo"
+
+    updating_reloaded = read_hdf5(fname=test_file)
+
+    assert updating_reloaded is updating
+
+    # Because we manually defined __setstate__ to _update_ state data, we should get
+    # the saved `foo` and the post-save `bar` attribute after loading
+    assert updating_reloaded.foo == "foo"  # Load over-wrote "not foo"
+    assert updating.bar == "bar"
+    assert updating_reloaded.bar == "bar"
```

Comment:
I am not sure if this part is necessary. As far as I understand it is a demonstration how to work around the issue that data which is added between saving and loading is lost. Intuitively the solution is to define a custom `__reduce__()` method. I feel a comment would be sufficient to explain this, it is not necessary to repeat the test.